nomos-services/cryptarchia-consensus/src/lib.rs

```diff
@@ -872,6 +854,206 @@ where
+    ///
+    /// # Returns
+    ///
+    /// A vector of blocks in the range from `from` to `to`.
+    /// If no blocks are found, returns an empty vector.
+    /// If any of the HeaderId are invalid, returns an error with the first invalid header id.
+    async fn get_blocks_in_range(
+        from: HeaderId,
+        to: HeaderId,
+        storage_adapter: &StorageAdapter<Storage, TxS::Tx, BS::BlobId>,
+    ) -> Vec<Block<ClPool::Item, DaPool::Item>> {
+        // Due to the blocks traversal order, this yields `to..from` order
+        let blocks = futures::stream::unfold(to, |header_id| async move {
+            if header_id == from {
+                None
+            } else {
+                let block = storage_adapter
+                    .get_block(&header_id)
+                    .await
+                    .unwrap_or_else(|| {
+                        panic!("Could not retrieve block {to} from storage during recovery")
+                    });
+                let parent_header_id = block.header().parent();
+                Some((block, parent_header_id))
+            }
+        });
+
+        // To avoid confusion, the order is reversed so it fits the natural `from..to` order
+        blocks.collect::<Vec<_>>().await.into_iter().rev().collect()
+    }
+
+    /// Builds cryptarchia
+    /// The build process is determined by the initial state passed:
+    /// - If the initial state doesn't contain any recovery information, cryptarchia is built from
+    ///     genesis.
+    /// - If it does, the recovery process is started: Recovery is done from genesis or security
+    ///     depending on the available information.
+    ///
+    /// # Arguments
+    ///
+    /// * `initial_state` - The initial state of cryptarchia.
+    /// * `genesis_id` - The genesis block id.
+    /// * `genesis_state` - The genesis ledger state.
+    /// * `leader` - The leader instance. It needs to be a Leader initialised to genesis. This
+    ///     function will update the leader if needed.
+    /// * `ledger_config` - The ledger configuration.
+    /// * `relays` - The relays object containing all the necessary relays for the consensus.
+    /// * `block_subscription_sender` - The broadcast channel to send the blocks to the services.
+    async fn build_cryptarchia(
+        mut initial_state: CryptarchiaConsensusState<
+            TxS::Settings,
+            BS::Settings,
+            NetAdapter::Settings,
+            BlendAdapter::Settings,
+        >,
+        genesis_id: HeaderId,
+        genesis_state: LedgerState,
+        ledger_config: nomos_ledger::Config,
+        leader_config: LeaderConfig,
+        relays: &CryptarchiaConsensusRelays<
+            BlendAdapter,
+            BS,
+            ClPool,
+            ClPoolAdapter,
+            DaPool,
+            DaPoolAdapter,
+            NetAdapter,
+            SamplingBackend,
+            SamplingRng,
+            Storage,
+            TxS,
+        >,
```

Comment:
Now that I see this. CryptarchiaConsensusRelays should be a type alias to hide generics except in a single point.

Reply:
Is there a way to do it without an associated type?